File: test/ocaml/opcode/test_op_references.ml
open Core.Std
open OUnit2
open Jvalue
open TestEnv

let jclass = compile_jclass "test/java/opcode/TestOpReferences.java"

let run ?(cmp = (=)) method_name method_descriptor result =
  let actual = run_method jclass method_name method_descriptor in
  assert_equal ~cmp actual (Some result)

let suite =
  "op_references" >:::
  [ "op_getstatic" >:: (fun _ -> run "op_getstatic" "()I" (Int 0l));
    "op_putstatic" >:: (fun _ -> run "op_putstatic" "()I" (Int 2l));
    "op_invokestatic" >:: (fun _ -> run "op_invokestatic" "()J" (Long 699359257085L));
  ]

let () =
  run_test_tt_main suite

File: test/java/opcode/TestOpReferences.java
public class TestOpReferences {
  static int i;

  // static final String str = "Hello";

  static long foo (long z, int p, long q) {
    int i = 10 * p;
    long r = i * 1024 + q;
    return r * z;
  }

  static {
    i = 100;
    i = i * 10;
  }

  int op_getstatic () {
    return i;
  }

  int op_putstatic () {
    i = 2;
    return i;
  }

  static long op_invokestatic () {
    int p = 5;
    long i = foo (6666, 1024, 5678);
    long j = i * 10 + p;
    return j;
  }

  public static void main(String[] args) {
    System.out.println(op_invokestatic());
  }
}
